File: src/test/java/rectangleIntersection/helpfulClasses/CoordinateRectangleTest.java
package rectangleIntersection.helpfulClasses;

import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class CoordinateRectangleTest {

    @Test
    public void bottom_left_and_top_right_results_in_a_valid_rectangle() {
        CoordinateRectangle coordinateRectangle = new CoordinateRectangle(
                new Coordinates(0, 0),
                new Coordinates(1, 1)
        );

        assertThat(coordinateRectangle.getBottomLeft().getX()).isEqualTo(0);
        assertThat(coordinateRectangle.getBottomLeft().getY()).isEqualTo(0);

        assertThat(coordinateRectangle.getTopRight().getX()).isEqualTo(1);
        assertThat(coordinateRectangle.getTopRight().getY()).isEqualTo(1);
    }

    @Test
    public void bottom_right_and_top_left_results_in_a_valid_rectangle() {
        CoordinateRectangle coordinateRectangle = new CoordinateRectangle(
                new Coordinates(1, 0),
                new Coordinates(0, 1)
        );

        assertThat(coordinateRectangle.getBottomLeft().getX()).isEqualTo(0);
        assertThat(coordinateRectangle.getBottomLeft().getY()).isEqualTo(0);

        assertThat(coordinateRectangle.getTopRight().getX()).isEqualTo(1);
        assertThat(coordinateRectangle.getTopRight().getY()).isEqualTo(1);
    }

    @Test(expected = InvalidRectangle.class)
    public void if_the_rectangle_is_a_vertical_line_throws_an_exception() {
        Integer x = 0;

        new CoordinateRectangle(
                new Coordinates(x, 0),
                new Coordinates(x, 100)
        );
    }

    @Test(expected = InvalidRectangle.class)
    public void if_the_rectangle_is_a_horizontal_line_throws_an_exception() {
        Integer y = 0;

        new CoordinateRectangle(
                new Coordinates(0, y),
                new Coordinates(100, y)
        );
    }
}

File: src/test/kotlin/kotlinCoreTDDTests/rectangleIntersection/helpfulClasses/CoordinateRectangleTest.kt
package kotlinCoreTDDTests.rectangleIntersection.helpfulClasses

import kotlinCoreTDD.rectangleIntersection.helpfulClasses.InvalidRectangle
import kotlinCoreTDD.rectangleIntersection.helpfulClasses.Coordinate
import kotlinCoreTDD.rectangleIntersection.helpfulClasses.CoordinateRectangle
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class CoordinateRectangleTest {

    @Test(expected = InvalidRectangle::class)
    fun error_if_thereIsNoSpaceBetweenXs() {
        CoordinateRectangle(
                Coordinate(x = 0, y = 1),
                Coordinate(x = 0, y = 2)
        )
    }

    @Test(expected = InvalidRectangle::class)
    fun error_if_thereIsNoSpaceBetweenYs() {
        CoordinateRectangle(
                Coordinate(x = 0, y = 1),
                Coordinate(x = 1, y = 1)
        )
    }

    @Test
    fun smallXSmallYCoordinateHasTheSmallestXandY() {
        listOf(
                Pair(
                        CoordinateRectangle(
                                Coordinate(x = 0, y = 1),
                                Coordinate(x = 1, y = 7)
                        ),
                        Coordinate(x = 0, y = 1)
                ),
                Pair(
                        CoordinateRectangle(
                                Coordinate(x = 1, y = 7),
                                Coordinate(x = 2, y = 1)
                        ),
                        Coordinate(x = 1, y = 1)
                ),
                Pair(
                        CoordinateRectangle(
                                Coordinate(x = 2, y = 7),
                                Coordinate(x = 1, y = 1)
                        ),
                        Coordinate(x = 1, y = 1)
                ),
                Pair(
                        CoordinateRectangle(
                                Coordinate(x = 3, y = 2),
                                Coordinate(x = 2, y = 7)
                        ),
                        Coordinate(x = 2, y = 2)
                )
        ).forEach { pair ->
            assertThat(pair.first.smallXSmallY()).isEqualTo(pair.second)
        }
    }

    @Test
    fun bigXBigYCoordinateHasTheLargestXandY() {
        listOf(
                Pair(
                        CoordinateRectangle(
                                Coordinate(x = 0, y = 1),
                                Coordinate(x = 1, y = 7)
                        ),
                        Coordinate(x = 1, y = 7)
                ),
                Pair(
                        CoordinateRectangle(
                                Coordinate(x = 1, y = 7),
                                Coordinate(x = 2, y = 1)
                        ),
                        Coordinate(x = 2, y = 7)
                ),
                Pair(
                        CoordinateRectangle(
                                Coordinate(x = 2, y = 9),
                                Coordinate(x = 1, y = 1)
                        ),
                        Coordinate(x = 2, y = 9)
                ),
                Pair(
                        CoordinateRectangle(
                                Coordinate(x = 3, y = 2),
                                Coordinate(x = 2, y = 7)
                        ),
                        Coordinate(x = 3, y = 7)
                )
        ).forEach { pair ->
            assertThat(pair.first.bigXBigY()).isEqualTo(pair.second)
        }
    }
}